### ISSUE TITLE
[bug] The Crash.

### ISSUE THREAD

**Issue by RowanDDR** [bazzinotti/snestracker#160]
Sometimes SNES Tracker "crashes" .. or to be more precise, it locks up (becomes unresponsive to input). On my system (Win 10 64bit), 99.99% of the time it comes back to life after about 10 seconds, following a lockup. I don't think I've ever lost any work to a crash. So if it happens to you, don't panic, just wait.

I've observed that something that usually happens a in the ~5 seconds BEFORE an ST crash is that when inputting notes, there is suddenly no sound (preview broken).. that's a surefire sign that the lockup crash is coming in a few seconds time. BUT... sometimes the preview is broken and the crash does _not_ happen.. the preview then fixes itself after pressing return (starting the song playing).

Also, the crash usually happens when I'm doing "something". i.e. I've never seen it crash when idle, or when a song is playing and I'm sat back.

The lockup usually occurs when I'm inputting or previewing notes on my midi keyboard at a moderately rapid pace... I will try only inputting via the PC keyboard to see if that narrows it down.

If anyone else can think of ways to reproduce the crash please mention it here. Let's try and narrow it down.

For me this has affected every version of SNES Tracker up to and including v0.2.1.

![image](https://user-images.githubusercontent.com/6626459/98187199-4b645c80-1f08-11eb-97ec-234fbaaa7577.png)

**Comment by bazzinotti** [comment]
Rowan, thanks for your wonderful feedback!

Note to self: I believe this bug is not present in v0.1.0, which was before note playback was implemented. It might be worth checking.

If we could only find a reliable way to trigger this bug immediately after starting the program, that would greatly help!!

I will play around with this when I get home. I really want to fix this one!! Wish me luck!

**Comment by bazzinotti** [comment]
tldr; I feel much better about being able to fix this bug. Here are some notes from messing around with it briefly:

The only way I know to cause the bug:

1. Load a song with at least one instrument loaded (this is merely to provide audio confirmation that the bug has occurred).
2. input notes as fast as I can, whether in playback or edit mode.
3. When the notes no longer create sound, the bug has been triggered. At this point, You will no longer hear any notes played back. However the app will continue to function until it has the Music Player call SDL_PauseAudioDevice(), typically by pressing Return to play the song.

- I used my PC keyboard, so that rules out it being a MIDI issue.
- Sometimes I cannot trigger the bug no matter how long I try. In that case, I restart the app and I can usually get it within 2-3 restarts.
- I've confirmed it is not necessary to play song to cause this bug.

Here's a stack trace AFTER the bug has already been triggered and the eventual  call to SDL_PauseAudioDevice() freezes the app for about 15 seconds.

```
render_to_apu(); playback: 1
c^C
Thread 1 "st" received signal SIGINT, Interrupt.
__lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:103
103	../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S: No such file or directory.
(gdb) bt
#0  0x00007ffff7437c7c in __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:103
#1  0x00007ffff7430274 in __GI___pthread_mutex_lock (mutex=0x55555578f5f0) at ../nptl/pthread_mutex_lock.c:115
#2  0x00007ffff739d93e in SDL_LockMutex_REAL (mutex=<optimized out>)
    at /home/bazz/Downloads/SDL2-2.0.10/src/thread/pthread/SDL_sysmutex.c:108
#3  0x00007ffff72af2e9 in SDL_PauseAudioDevice_REAL (devid=<optimized out>, pause_on=0)
    at /home/bazz/Downloads/SDL2-2.0.10/src/audio/SDL_audio.c:1532
#4  0x00005555555808b8 in sound_start() () at shared/gme/player/Music_Player.cpp:632
#5  0x000055555557fd93 in Music_Player::pause(int, bool, bool) (this=0x55555575e190, b=0, with_fade=false, fade_threaded=false)
    at shared/gme/player/Music_Player.cpp:372
#6  0x0000555555583398 in Tracker::render_to_apu(bool, bool) (this=0x7ffff0094010, repeat_pattern=false, startFromPlayhead=false)
    at tracker/Tracker.cpp:833
#7  0x000055555558249b in Tracker::handle_events() (this=0x7ffff0094010) at tracker/Tracker.cpp:525
#8  0x00005555555815c2 in Tracker::run() (this=0x7ffff0094010) at tracker/Tracker.cpp:146
#9  0x000055555559a85a in App::run() (this=0x7fffffffd6b0) at tracker/app.cpp:157
#10 0x000055555559923f in main(int, char**) (argc=1, argv=0x7fffffffd7f8) at tracker/main.cpp:49
(gdb) r
```

**Comment by bazzinotti** [comment]
I think I've fixed this bug!! The following text is a note to self.

I narrowed the problem down to libgme_m's Spc_Dsp::run()'s clocks_remain loop variable having a SUPER LARGE value, effectively causing the 15-second crash. That clocks_remain is originally set from a call to run_until_()

Somehow, I noticed run_until_() being called from Snes_Spc::read_port(), and that I seemed to have removed it for a certain reason from write_port() in the past. Removing that call seems to have done the trick, and I'm not seeing any adverse effects (yet). 

For reference, here's a call stack from the audio thread at the same point in time as the main thread's call stack previously posted above.

```
#0  0x00007ffff7190c7e in Spc_Dsp::run(int) (this=this@entry=0x5555559439e0, clocks_remain=2096230256) at gme_m/Spc_Dsp.cpp:889
#1  0x00007ffff71960a3 in Snes_Spc::dsp_write(int, int) (time=<optimized out>, data=127, this=0x5555559439e0) at gme_m/Snes_Spc.cpp:164
#2  0x00007ffff71960a3 in Snes_Spc::cpu_write_smp_reg(int, int, int)
    (this=0x5555559439e0, data=127, time=<optimized out>, addr=<optimized out>) at gme_m/Snes_Spc.cpp:400
#3  0x00007ffff7196a39 in Snes_Spc::run_until_(int) (this=0x5555559439e0, end_time=<optimized out>) at gme_m/Spc_Cpu.h:450
#4  0x00007ffff7198d8d in Snes_Spc::end_frame(int) (this=this@entry=0x5555559439e0, end_time=end_time@entry=1024) at gme_m/Snes_Spc.cpp:564
#5  0x00007ffff71936c7 in Snes_Spc::play(int, short*) (this=0x5555559439e0, count=64, out=<optimized out>) at gme_m/Snes_Spc_misc.cpp:343
#6  0x00007ffff719508e in Music_Emu::emu_play(long, short*) (out=<optimized out>, count=<optimized out>, this=0x555555943220)
    at /usr/include/bits/string_fortified.h:71
#7  0x00007ffff719508e in Music_Emu::emu_play(long, short*) (this=0x555555943220, count=<optimized out>, out=<optimized out>)
    at gme_m/Music_Emu.cpp:290
#8  0x00007ffff7195483 in Music_Emu::play(long, short*) (this=0x555555943220, out_count=64, out=0x5555557b2ec0) at gme_m/Music_Emu.cpp:379
#9  0x0000555555580584 in Music_Player::fill_buffer(void*, short*, int) (data=0x555555789900, out=0x5555557b2ec0, count=64)
    at shared/gme/player/Music_Player.cpp:527
#10 0x00005555555806f5 in sdl_callback(void*, Uint8*, int) (data=0x0, out=0x5555557b2ec0 "\032", count=128)
    at shared/gme/player/Music_Player.cpp:570
#11 0x00007ffff72ac9bb in SDL_RunAudio (devicep=devicep@entry=0x5555557a0f20) at /home/bazz/Downloads/SDL2-2.0.10/src/audio/SDL_audio.c:735
#12 0x00007ffff7313fbc in SDL_RunThread (data=0x5555557a82b0) at /home/bazz/Downloads/SDL2-2.0.10/src/thread/SDL_thread.c:283
#13 0x00007ffff739d2d9 in RunThread (data=<optimized out>) at /home/bazz/Downloads/SDL2-2.0.10/src/thread/pthread/SDL_systhread.c:79
#14 0x00007ffff742d3a7 in start_thread (arg=<optimized out>) at pthread_create.c:486
#15 0x00007ffff6cdf77f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
